src/garage/_environment.py

```diff
@@ -0,0 +1,247 @@
+    +-----------------------+
+
+    Set the following properties:
+
+    +-----------------------+-------------------------------------------------+
+    | Properties            | Description                                     |
+    +=======================+=================================================+
+    | action_space          | The action space specification                  |
+    +-----------------------+-------------------------------------------------+
+    | observation_space     | The observation space specification             |
+    +-----------------------+-------------------------------------------------+
+    | spec                  | The environment specifications                  |
+    +-----------------------+-------------------------------------------------+
+    | render_modes          | The list of supported render modes              |
+    +-----------------------+-------------------------------------------------+
+    | max_episode_steps     | The maximum number of steps (i.e. time          |
+    |                       | limit) of the environment.                      |
+    +-----------------------+-------------------------------------------------+
+
+    Make sure your environment is pickle-able:
+        Garage pickles the environment via the `cloudpickle` module
+        to save snapshots of the experiment. However, some environments may
+        contain attributes that are not pickle-able (e.g. a client-server
+        connection). In such cases, override `__setstate__()` and
+        `__getstate()__` to add your custom pickle logic.
+
+        You might want to refer to the EzPickle module:
+        https://github.com/openai/gym/blob/master/gym/utils/ezpickle.py
+        for a lightweight way of pickle and unpickle via constructor
+        arguments.
+
+    """
+
+    @property
+    @abc.abstractmethod
+    def action_space(self):
+        """Return the action space specification.
+
+        Some environments do not provide their action space specifications as
+        type akro.Space. In such cases, subclasses should silently convert
+        the environment's action_space to akro.Space.
+
+        Returns:
+            akro.Space: The environment specification.
+
+        """
+
+    @property
+    @abc.abstractmethod
+    def observation_space(self):
+        """Return the observation space specification.
+
+        Some environments do not provide their observation space
+        specifications as type akro.Space. In such cases, subclasses should
+        silently convert the environment's observation_space to akro.Space.
+
+        Returns:
+            akro.Space: The environment specification.
+        """
+
+    @property
+    @abc.abstractmethod
+    def spec(self):
+        """Return the environment specification.
+
+        Returns:
+            garage.EnvSpec: The environment specification.
+
+        """
+
+    @property
+    @abc.abstractmethod
+    def max_episode_steps(self):
```

Comment:
I wonder if this should be part of `EnvSpec` instead? It's often needed in similar places.

Reply:
i agree -- it's an important part of the EnvSpec because it's something you are likely to need to know to construct a function approximator associated with this environment.